distributed/scheduler.py

```diff
@@ -657,134 +778,277 @@ class TaskState:
     :      The group of tasks to which this one belongs.
     """
 
+    _key: str
+    _hash: Py_hash_t
+    _prefix: object
+    _run_spec: object
+    _priority: tuple
+    _state: str
```

Comment:
Should we make this into an enum?  Would Cython prefer that?

Reply:
Cython can't do much with a Python `Enum`. There are `enum`s in Cython that it does expose to the Python layer optionally. Though these are not available in pure Python mode unfortunately ( https://github.com/cython/cython/issues/3923 ).

For now I think it is ok to just type this as a `str` as part of our broad effort to type things. We can then revisit specifics once everything is in and we have had an opportunity to profile more closely.

Reply:
Sounds good.  We check and change state pretty often.  I wouldn't be surprised if this has some effect in the future. 

Are pure Python enums something that would make sense to request upstream, or is this likely to be hard to achieve with cython?

Reply:
Yeah it may. I think line profiling after this lands makes sense and should hopefully guide us to which of several things we should further optimize. It could be `Status` or it could be something else.

Seems like a reasonable request to me (though maybe I'm biased as I already made that request 😄). Have no idea if it will be easy or not. Certainly hope it is doable.

In any event, I think at a first pass having `str` here is fine. Am more interested in broadly typing things first and worrying about more tuning in subsequent PRs after more profiling.

Reply:
One other thing I forgot to note, Cython does `intern` any `str` literals it sees by default as part of module initialization (IOW at `import` time), which it hangs onto globally (so they don't go out of scope). That way, when we do something like `ts._state = "running"`, Cython will already have `intern`ed `"running"` and assigned the `intern`ed `str` variable to `ts._state`. Further when we later do something like `assert ts._state == "running"`, this `"running"` will also refer to the same `intern`ed `str`. As a result comparisons of `str`s defined with literals should benefit from the `intern` speedup.

Though there is a caveat. This does not apply for any dynamically generated `str`s. So `"My name is %s" % s` will construct a `str` at runtime that is not `intern`ed.